File: svg_editor.js
// svg_editor.js
// Версия на JavaScript с использованием xml2js, commander, fs

const fs = require('fs');
const { Command } = require('commander');
const xml2js = require('xml2js');

class SvgEditor {
    constructor(filepath) {
        this.filepath = filepath;
        this.xml = null;
        this.svg = null;
        this.layers = [];
    }

    async load() {
        const data = await fs.promises.readFile(this.filepath, 'utf-8');
        const parser = new xml2js.Parser({ explicitArray: false });
        this.xml = await parser.parseStringPromise(data);
        this.svg = this.xml.svg;
        this._collectLayers();
    }

    _collectLayers() {
        this.layers = [];
        if (!this.svg) return;
        const children = this.svg.g || [];
        // если только один слой, xml2js может сделать его объектом, а не массивом
        const items = Array.isArray(children) ? children : [children];
        items.forEach((elem, idx) => {
            // группа может быть без явного тега, но мы ищем по $ (атрибуты)
            if (elem.$ && elem.$.id) {
                this.layers.push({ index: idx, element: elem, id: elem.$.id, attrs: elem.$ });
            } else if (elem.$ && elem.$['data-layer']) {
                this.layers.push({ index: idx, element: elem, id: elem.$['data-layer'], attrs: elem.$ });
            } else {
                // считаем любые элементы верхнего уровня слоями, если есть атрибут data-layer
                this.layers.push({ index: idx, element: elem, id: elem.$?.['data-layer'] || null, attrs: elem.$ || {} });
            }
        });
    }

    listLayers() {
        if (this.layers.length === 0) {
            console.log('Слои не найдены.');
            return;
        }
        this.layers.forEach(l => {
            const attrsStr = Object.entries(l.attrs).map(([k,v]) => `${k}="${v}"`).join(' ');
            console.log(`[${l.index}] ${l.id || 'без id'} : ${attrsStr}`);
        });
    }

    addLayer(id, extraAttrs = {}) {
        const newGroup = { $: { id, ...extraAttrs } };
        if (!this.svg.g) {
            this.svg.g = [];
        }
        // если это не массив, преобразуем
        if (!Array.isArray(this.svg.g)) {
            this.svg.g = [this.svg.g];
        }
        this.svg.g.push(newGroup);
        this._collectLayers(); // перезагружаем список
        console.log(`Слой '${id}' добавлен.`);
    }

    removeLayer(index) {
        if (index < 0 || index >= this.layers.length) {
            throw new Error('Индекс вне диапазона.');
        }
        const layers = this.svg.g;
        if (Array.isArray(layers)) {
            layers.splice(index, 1);
        } else if (layers) {
            // если только один элемент, удаляем весь массив
            this.svg.g = null;
        }
        this._collectLayers();
        console.log(`Слой ${index} удалён.`);
    }

    moveLayer(index, direction) {
        if (index < 0 || index >= this.layers.length) {
            throw new Error('Индекс вне диапазона.');
        }
        const layers = this.svg.g;
        if (!Array.isArray(layers)) {
            throw new Error('Невозможно переместить: не массив слоёв.');
        }
        let newIndex = direction === 'up' ? index - 1 : index + 1;
        if (newIndex < 0 || newIndex >= layers.length) {
            throw new Error('Невозможно переместить в указанном направлении.');
        }
        // меняем местами
        [layers[index], layers[newIndex]] = [layers[newIndex], layers[index]];
        this._collectLayers();
        console.log(`Слой ${index} перемещён ${direction === 'up' ? 'вверх' : 'вниз'}.`);
    }

    editLayer(index, attr, value) {
        if (index < 0 || index >= this.layers.length) {
            throw new Error('Слой не найден.');
        }
        const layer = this.layers[index];
        if (!layer.attrs) layer.attrs = {};
        layer.attrs[attr] = value;
        // обновляем элемент
        if (layer.element.$) {
            layer.element.$[attr] = value;
        } else {
            layer.element.$ = { [attr]: value };
        }
        console.log(`Атрибут '${attr}' слоя ${index} установлен в '${value}'.`);
    }

    async save(outputFile) {
        const builder = new xml2js.Builder({ xmldec: { version: '1.0', encoding: 'UTF-8' } });
        const xmlString = builder.buildObject(this.xml);
        const out = outputFile || this.filepath;
        await fs.promises.writeFile(out, xmlString, 'utf-8');
        console.log(`Сохранено в ${out}`);
    }
}

// CLI с commander
const program = new Command();
program
    .name('svg_editor')
    .description('SVG Layer Editor (JavaScript)')
    .argument('<input>', 'Входной SVG файл')
    .action((input) => {
        // сохранение input для использования в подкомандах
        program.input = input;
    });

program
    .command('list')
    .description('Показать слои')
    .action(async () => {
        const editor = new SvgEditor(program.input);
        await editor.load();
        editor.listLayers();
    });

program
    .command('add-layer')
    .description('Добавить слой')
    .requiredOption('--id <id>', 'ID слоя')
    .option('--attr <key=value>', 'Дополнительные атрибуты', (val, prev) => {
        // собираем атрибуты в объект
        if (!prev) prev = {};
        const [key, value] = val.split('=');
        prev[key] = value;
        return prev;
    }, {})
    .action(async (options) => {
        const editor = new SvgEditor(program.input);
        await editor.load();
        editor.addLayer(options.id, options.attr);
        await editor.save(program.input); // сохраняем изменения
    });

program
    .command('remove-layer')
    .description('Удалить слой')
    .requiredOption('--index <n>', 'Индекс слоя', parseInt)
    .action(async (options) => {
        const editor = new SvgEditor(program.input);
        await editor.load();
        editor.removeLayer(options.index);
        await editor.save(program.input);
    });

program
    .command('move-layer')
    .description('Переместить слой')
    .requiredOption('--index <n>', 'Индекс слоя', parseInt)
    .requiredOption('--direction <up|down>', 'Направление')
    .action(async (options) => {
        const editor = new SvgEditor(program.input);
        await editor.load();
        editor.moveLayer(options.index, options.direction);
        await editor.save(program.input);
    });

program
    .command('edit-layer')
    .description('Изменить атрибут слоя')
    .requiredOption('--index <n>', 'Индекс слоя', parseInt)
    .requiredOption('--attr <key>', 'Имя атрибута')
    .requiredOption('--value <val>', 'Новое значение')
    .action(async (options) => {
        const editor = new SvgEditor(program.input);
        await editor.load();
        editor.editLayer(options.index, options.attr, options.value);
        await editor.save(program.input);
    });

program
    .command('save')
    .description('Сохранить SVG')
    .option('--output <file>', 'Выходной файл')
    .action(async (options) => {
        // если команда save вызвана без изменений, просто копируем
        const editor = new SvgEditor(program.input);
        await editor.load();
        await editor.save(options.output);
    });

program.parse(process.argv);

File: svg_editor.go
// svg_editor.go
// Версия на Go с использованием encoding/xml, горутинами для параллельного парсинга (опционально)

package main

import (
	"encoding/xml"
	"flag"
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

// SVG структуры для парсинга
type SVG struct {
	XMLName xml.Name `xml:"svg"`
	Groups  []Group  `xml:"g"`
	// другие элементы игнорируем
}

type Group struct {
	XMLName xml.Name `xml:"g"`
	Attrs   []xml.Attr `xml:",any,attr"`
	// содержимое игнорируем для простоты
}

type Layer struct {
	Index   int
	Element Group
	ID      string
}

type Editor struct {
	filename string
	svg      SVG
	layers   []Layer
}

func NewEditor(filename string) (*Editor, error) {
	data, err := ioutil.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	var svg SVG
	err = xml.Unmarshal(data, &svg)
	if err != nil {
		return nil, err
	}
	e := &Editor{filename: filename, svg: svg}
	e.collectLayers()
	return e, nil
}

func (e *Editor) collectLayers() {
	e.layers = []Layer{}
	for i, g := range e.svg.Groups {
		id := ""
		for _, attr := range g.Attrs {
			if attr.Name.Local == "id" {
				id = attr.Value
				break
			}
			if attr.Name.Local == "data-layer" {
				id = attr.Value
			}
		}
		e.layers = append(e.layers, Layer{Index: i, Element: g, ID: id})
	}
}

func (e *Editor) ListLayers() {
	if len(e.layers) == 0 {
		fmt.Println("Слои не найдены.")
		return
	}
	for _, l := range e.layers {
		attrs := ""
		for _, a := range l.Element.Attrs {
			attrs += fmt.Sprintf(`%s="%s" `, a.Name.Local, a.Value)
		}
		fmt.Printf("[%d] %s : %s\n", l.Index, l.ID, attrs)
	}
}

func (e *Editor) AddLayer(id string, extraAttrs map[string]string) {
	newGroup := Group{}
	attrs := []xml.Attr{}
	if id != "" {
		attrs = append(attrs, xml.Attr{Name: xml.Name{Local: "id"}, Value: id})
	}
	for k, v := range extraAttrs {
		attrs = append(attrs, xml.Attr{Name: xml.Name{Local: k}, Value: v})
	}
	newGroup.Attrs = attrs
	e.svg.Groups = append(e.svg.Groups, newGroup)
	e.collectLayers()
	fmt.Printf("Слой '%s' добавлен.\n", id)
}

func (e *Editor) RemoveLayer(index int) {
	if index < 0 || index >= len(e.svg.Groups) {
		panic("Индекс вне диапазона")
	}
	e.svg.Groups = append(e.svg.Groups[:index], e.svg.Groups[index+1:]...)
	e.collectLayers()
	fmt.Printf("Слой %d удалён.\n", index)
}

func (e *Editor) MoveLayer(index int, direction string) {
	if index < 0 || index >= len(e.svg.Groups) {
		panic("Индекс вне диапазона")
	}
	newIdx := index - 1
	if direction == "down" {
		newIdx = index + 1
	}
	if newIdx < 0 || newIdx >= len(e.svg.Groups) {
		panic("Невозможно переместить в указанном направлении")
	}
	e.svg.Groups[index], e.svg.Groups[newIdx] = e.svg.Groups[newIdx], e.svg.Groups[index]
	e.collectLayers()
	fmt.Printf("Слой %d перемещён %s.\n", index, direction)
}

func (e *Editor) EditLayer(index int, attr, value string) {
	if index < 0 || index >= len(e.svg.Groups) {
		panic("Слой не найден")
	}
	// ищем атрибут
	found := false
	for i := range e.svg.Groups[index].Attrs {
		if e.svg.Groups[index].Attrs[i].Name.Local == attr {
			e.svg.Groups[index].Attrs[i].Value = value
			found = true
			break
		}
	}
	if !found {
		e.svg.Groups[index].Attrs = append(e.svg.Groups[index].Attrs, xml.Attr{Name: xml.Name{Local: attr}, Value: value})
	}
	e.collectLayers()
	fmt.Printf("Атрибут '%s' слоя %d установлен в '%s'.\n", attr, index, value)
}

func (e *Editor) Save(output string) error {
	out := output
	if out == "" {
		out = e.filename
	}
	data, err := xml.MarshalIndent(e.svg, "", "  ")
	if err != nil {
		return err
	}
	// добавляем заголовок XML
	data = []byte(xml.Header + string(data))
	return ioutil.WriteFile(out, data, 0644)
}

func main() {
	// Простой парсинг флагов: используем подкоманды через os.Args
	if len(os.Args) < 3 {
		fmt.Println("Использование: go run svg_editor.go <input.svg> <команда> [параметры]")
		os.Exit(1)
	}
	input := os.Args[1]
	command := os.Args[2]

	editor, err := NewEditor(input)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Ошибка загрузки: %v\n", err)
		os.Exit(1)
	}

	switch command {
	case "list":
		editor.ListLayers()
	case "add-layer":
		// args: --id <id> [--attr key=value ...]
		id := ""
		attrs := map[string]string{}
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--id" && i+1 < len(os.Args) {
				id = os.Args[i+1]
				i++
			} else if os.Args[i] == "--attr" && i+1 < len(os.Args) {
				kv := strings.SplitN(os.Args[i+1], "=", 2)
				if len(kv) == 2 {
					attrs[kv[0]] = kv[1]
				}
				i++
			}
		}
		editor.AddLayer(id, attrs)
		editor.Save("")
	case "remove-layer":
		// --index N
		index := -1
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--index" && i+1 < len(os.Args) {
				idx, err := strconv.Atoi(os.Args[i+1])
				if err == nil {
					index = idx
				}
				i++
			}
		}
		if index < 0 {
			fmt.Println("Не указан --index")
			os.Exit(1)
		}
		editor.RemoveLayer(index)
		editor.Save("")
	case "move-layer":
		index := -1
		direction := ""
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--index" && i+1 < len(os.Args) {
				idx, err := strconv.Atoi(os.Args[i+1])
				if err == nil {
					index = idx
				}
				i++
			} else if os.Args[i] == "--direction" && i+1 < len(os.Args) {
				direction = os.Args[i+1]
				i++
			}
		}
		if index < 0 || direction == "" {
			fmt.Println("Не указаны --index или --direction")
			os.Exit(1)
		}
		editor.MoveLayer(index, direction)
		editor.Save("")
	case "edit-layer":
		index := -1
		attr := ""
		value := ""
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--index" && i+1 < len(os.Args) {
				idx, err := strconv.Atoi(os.Args[i+1])
				if err == nil {
					index = idx
				}
				i++
			} else if os.Args[i] == "--attr" && i+1 < len(os.Args) {
				attr = os.Args[i+1]
				i++
			} else if os.Args[i] == "--value" && i+1 < len(os.Args) {
				value = os.Args[i+1]
				i++
			}
		}
		if index < 0 || attr == "" || value == "" {
			fmt.Println("Не указаны --index, --attr, --value")
			os.Exit(1)
		}
		editor.EditLayer(index, attr, value)
		editor.Save("")
	case "save":
		output := ""
		for i := 3; i < len(os.Args); i++ {
			if os.Args[i] == "--output" && i+1 < len(os.Args) {
				output = os.Args[i+1]
				i++
			}
		}
		editor.Save(output)
	default:
		fmt.Printf("Неизвестная команда: %s\n", command)
		os.Exit(1)
	}
}
